pkg/liquidity-source/fluid/dex-t1/pool_simulator.go

```diff
@@ -0,0 +1,547 @@
+ * @param {number} iReserveIn - Imaginary token reserve with input amount.
+ * @param {number} iReserveOut - Imaginary token reserve of output amount.
+ * @returns {number} - The maximum output amount of the other asset.
+ */
+func getAmountOut(amountIn *big.Int, iReserveIn *big.Int, iReserveOut *big.Int) *big.Int {
+	// Both numerator and denominator are scaled to 1e6 to factor in fee scaling.
+	numerator := new(big.Int).Mul(amountIn, iReserveOut)
+	denominator := new(big.Int).Add(iReserveIn, amountIn)
+
+	// Using the swap formula: (AmountIn * iReserveY) / (iReserveX + AmountIn)
+	return new(big.Int).Div(numerator, denominator)
+}
+
+/**
+ * Given an output amount of asset and pair reserves, returns the input amount of the other asset
+ * @param {number} amountOut - Desired output amount of the asset.
+ * @param {number} iReserveIn - Imaginary token reserve of input amount.
+ * @param {number} iReserveOut - Imaginary token reserve of output amount.
+ * @returns {number} - The input amount of the other asset.
+ */
+func getAmountIn(amountOut *big.Int, iReserveIn *big.Int, iReserveOut *big.Int) *big.Int {
+	// Both numerator and denominator are scaled to 1e6 to factor in fee scaling.
+	numerator := new(big.Int).Mul(amountOut, iReserveIn)
+	denominator := new(big.Int).Sub(iReserveOut, amountOut)
+
+	// Using the swap formula: (AmountOut * iReserveX) / (iReserveY - AmountOut)
+	return new(big.Int).Div(numerator, denominator)
+}
+
+/**
+ * Calculates how much of a swap should go through the collateral pool.
+ * @param {number} t - Total amount in.
+ * @param {number} x - Imaginary reserves of token out of collateral.
+ * @param {number} y - Imaginary reserves of token in of collateral.
+ * @param {number} x2 - Imaginary reserves of token out of debt.
+ * @param {number} y2 - Imaginary reserves of token in of debt.
+ * @returns {number} a - How much swap should go through collateral pool. Remaining will go from debt.
+ * @note If a < 0 then entire trade route through debt pool and debt pool arbitrage with col pool.
+ * @note If a > t then entire trade route through col pool and col pool arbitrage with debt pool.
+ * @note If a > 0 & a < t then swap will route through both pools.
+ */
+func swapRoutingIn(t *big.Int, x *big.Int, y *big.Int, x2 *big.Int, y2 *big.Int) *big.Int {
+	bI1e18 := new(big.Int)
+	bI1e18.SetString(String1e18, 10) // 1e18
+
+	xyProduct := new(big.Int).Mul(x, y)
+	xyProduct = new(big.Int).Mul(xyProduct, bI1e18)
+	xyRoot := new(big.Int).Sqrt(xyProduct)
+
+	x2y2Product := new(big.Int).Mul(x2, y2)
+	x2y2Product = new(big.Int).Mul(x2y2Product, bI1e18)
+	x2y2Root := new(big.Int).Sqrt(x2y2Product)
+
+	y2xyRoot := new(big.Int).Mul(y2, xyRoot)
+	txyRoot := new(big.Int).Mul(t, xyRoot)
+	yx2y2Root := new(big.Int).Mul(y, x2y2Root)
+	sum := new(big.Int).Add(y2xyRoot, txyRoot)
+	sum = new(big.Int).Sub(sum, yx2y2Root)
+	denominator := new(big.Int).Add(xyRoot, x2y2Root)
+	a := new(big.Int).Div(sum, denominator)
+	return a
```

Comment:
reduce and use local var for temporary numbers

```suggestion
// constant.go
var bI1e18, _ = new(big.Int).SetString(String1e18, 10) // 1e18

func swapRoutingIn(t *big.Int, x *big.Int, y *big.Int, x2 *big.Int, y2 *big.Int) *big.Int {
	var xyRoot, x2y2Root big.Int
	xyRoot.Mul(x, y).Mul(&xyRoot, bI1e18).Sqrt(&xyRoot)
	x2y2Root.Mul(x2, y2).Mul(&x2y2Root, bI1e18).Sqrt(&x2y2Root)

	var tmp big.Int
	numerator := new(big.Int)
	numerator.Mul(y2, &xyRoot).Add(numerator, tmp.Mul(t, &xyRoot)).Add(numerator, tmp.Mul(y, &x2y2Root))
	denominator := tmp.Add(&xyRoot, &x2y2Root)
	return numerator.Div(numerator, denominator)
```

Reply:
sure, done: https://github.com/KyberNetwork/kyberswap-dex-lib/pull/524/commits/80f81ae0c3678cf8d71d076e21057debf4520d6b